source/grid/tria.cc

```diff
@@ -2068,11 +2068,22 @@ namespace internal
             // need to allow that
             if (boundary_line->boundary_id != numbers::internal_face_boundary_id)
               {
-                AssertThrow (! (line->boundary_id() == numbers::internal_face_boundary_id),
-                             ExcInteriorLineCantBeBoundary(line->vertex_index(0),
-                                                           line->vertex_index(1),
-                                                           boundary_line->boundary_id));
-                line->set_boundary_id (boundary_line->boundary_id);
+                if (line->boundary_id() == numbers::internal_face_boundary_id)
+                  {
+                    if (boundary_line->manifold_id == numbers::flat_manifold_id)
+                      {
+                        AssertThrow (! (line->boundary_id() == numbers::internal_face_boundary_id),
```

Comment:
@bangerth , It must fail in this case. It should be replaced by `AssertThrow(false, Exc....)`, but this is to explicitly document what the `false` is. 

Here is a quick explanation of why this is so intricated...

The problem is that the default value for a `boundary_id` is `0`, so whenever you construct a `boundary_line` (which should probably be renamed simply `edge` here) to set an interior `manifold_id`, we don't know whether we should set its boundary_id to zero or to interior_boundary_id, because when reading the file the information of whether the line is at the boundary or at the interior is not there. Therefore we check if the boundary_line has `boundary_id` different from `numbers::internal_face_boundary_id`, but only throw an exception if also the `manifold_id` is set to its default value: this means that something wrong happened internally: you tried to read a mesh with interior ids, but did not specify that you wanted manifolds instead of boundary_ids.